### DISCUSSION TITLE
how to parse block content in lua script

### DISCUSSION THREAD

**Discussion by produnis** [quarto-dev/quarto-cli#8371] [Q&A] [answered]
### Description

Hi there,

I have a question about Lua scripting. Please take a look at my "Cheatsheet" extension on GitHub: [Cheatsheet Extension](https://github.com/produnis/quarto-cheatsheet). This extension produces a PDF file and introduces a filter where the following Quarto syntax:

```
:::{.cheat title="Info"}
foo bar
:::

```
is rendered to a TikZ picture node.

The issue I'm facing is that `pandoc.utils.stringify(block.content)` on [line 4 of my Lua script](https://github.com/produnis/quarto-cheatsheet/blob/main/_extensions/quarto-cheatsheet/cheatsheet.lua#L4) removes all formatting from `foo bar`. I need to use `pandoc.utils.stringify(block.content)` to prevent an error on [line 9](https://github.com/produnis/quarto-cheatsheet/blob/main/_extensions/quarto-cheatsheet/cheatsheet.lua#L9) (`bad argument #2 to 'concat' (table expected, got string`)).

Is there an alternative method to retain formatting?

As seen in [this example](https://github.com/produnis/quarto-cheatsheet/blob/main/git-Befehle.qmd), I currently have to resort to:

```
:::{.cheat title="Semester nach Abschluss in \\texttt{main} mergen"}
\\texttt{git checkout main}\\\\
\\texttt{git merge SoSe2024}
:::
```
to achieve proper output. However, this becomes cumbersome, especially when dealing with complex mathematical formulas.

Any suggestions or insights would be greatly appreciated!

**Comment by cderv** [comment, marked as answer]
`pandoc.utils.stringify()` is about transforming AST content to strings. As [doc](https://pandoc.org/lua-filters.html#pandoc.utils.stringify) says:
> Converts the given element (Pandoc, Meta, Block, or Inline) into a string **with all formatting removed**.

You either need to keep the `block.content` and add around it some AST content, then let Pandoc do the writing to the output format. Lua filters for Pandoc are all about modifying the document content in its normalized AST form, so you need to build around this. 

If you want to build a LaTeX  `pandoc.RawBlock()` in Lua from the `block.content`, you must directly convert it to LaTeX from within Lua filter. [`pandoc.write()`](https://pandoc.org/lua-filters.html#pandoc.write) lets you do that, providing the right options. 

Hope this helps clarify the behavior you are seeing.

**Reply by produnis** [reply]
thank you very much for your help. with `pandoc.write()`, it works now. Pushed a fix and it works now.
